flocker/node/agents/blockdevice.py

```diff
@@ -0,0 +1,340 @@
+    """
+    _attached_directory_name = 'attached'
+    _unattached_directory_name = 'unattached'
+
+    def __init__(self, root_path):
+        """
+        :param FilePath root_path: The path beneath which all loopback backing
+            files and their organising directories will be created.
+        """
+        self._root_path = root_path
+
+    @classmethod
+    def from_path(cls, root_path):
+        """
+        :param bytes root_path: The path to a directory in which loop back
+            backing files will be created. The directory is created if it does
+            not already exist.
+        :returns: A ``LoopbackBlockDeviceAPI`` with the supplied ``root_path``.
+        """
+        api = cls(root_path=FilePath(root_path))
+        api._initialise_directories()
+        return api
+
+    def _initialise_directories(self):
+        """
+        Create the root and sub-directories in which loopback files will be
+        created.
+        """
+        self._unattached_directory = self._root_path.child(
+            self._unattached_directory_name)
+
+        try:
+            self._unattached_directory.makedirs()
+        except OSError:
+            pass
+
+        self._attached_directory = self._root_path.child(
+            self._attached_directory_name)
+
+        try:
+            self._attached_directory.makedirs()
+        except OSError:
+            pass
+
+    def create_volume(self, size):
+        """
+        Create a "sparse" file of some size and put it in the ``unattached``
+        directory.
+
+        See ``IBlockDeviceAPI.create_volume`` for parameter and return type
+        documentation.
+        """
+        volume = BlockDeviceVolume(
+            blockdevice_id=unicode(uuid4()),
+            size=size,
+        )
+        with self._unattached_directory.child(
+            volume.blockdevice_id.encode('ascii')
+        ).open('wb') as f:
+            f.truncate(size)
+        return volume
+
+    def _get(self, blockdevice_id):
+        for volume in self.list_volumes():
+            if volume.blockdevice_id == blockdevice_id:
+                return volume
+        raise UnknownVolume(blockdevice_id)
+
+    def attach_volume(self, blockdevice_id, host):
+        """
+        Move an existing ``unattached`` file into a per-host directory and
+        create a loopback device backed by that file.
+
+        Note: Although `mkfs` can format files directly and `mount` can mount
+        files directly (with the `-o loop` option), we want to simulate a real
+        block device which will be allocated a real block device file on the
+        host to which it is attached. This allows the consumer of this API to
+        perform formatting and mount operations exactly the same as for a real
+        block device.
+
+        See ``IBlockDeviceAPI.attach_volume`` for parameter and return type
+        documentation.
+        """
+        volume = self._get(blockdevice_id)
+        if volume.host is None:
```

Comment:
Next time you do this - might be easier to read if logic was "if volume.host is not None: raise" and then rest of logic at function level. Not necessary to change in this case, though.